crates/subspace-networking/src/protocols/request_response/request_response_factory.rs

```diff
@@ -608,295 +616,323 @@ impl NetworkBehaviour for RequestResponseFactoryBehaviour {
+                protocol: protocol_name,
+                inner_channel,
+                response:
+                    OutgoingResponse {
+                        result: Ok(payload),
+                        ..
+                    },
+            }) = maybe_outcome
+            {
+                if let Some((protocol, _)) = self.protocols.get_mut(&*protocol_name) {
+                    if protocol.send_response(inner_channel, Ok(payload)).is_err() {
+                        // Note: Failure is handled further below when receiving
+                        // `InboundFailure` event from `RequestResponse` behaviour.
+                        debug!(
+                            target: LOG_TARGET,
+                            %request_id,
+                            "Failed to send response for request on protocol {} due to a \
+                            timeout or due to the connection to the peer being closed. \
+                            Dropping response",
+                            protocol_name,
+                        );
                     }
                 }
             }
 
-            for rq_rs_runner in &mut self.request_handlers {
-                // Future.Output == (), so we don't need a result here
-                let _ = rq_rs_runner.poll_unpin(cx);
+            // We didn't receive a Poll::Pending, so there might be more pending_responses ready.
+            needs_wakeup = true;
+        }
+
+        // We want to poll all the runners once, every time.
+        for rq_rs_runner in &mut self.request_handlers {
+            if let Poll::Ready(()) = rq_rs_runner.poll_unpin(cx) {
+                // We didn't receive a Poll::Pending, but runners are only expected to do work
+                // if there are pending events.
+                // TODO: work out if we need to schedule a manual wakeup here.
             }
+        }
 
-            // Poll request-responses protocols.
-            for (protocol, (behaviour, response_builder)) in &mut self.protocols {
-                while let Poll::Ready(event) = behaviour.poll(cx) {
-                    let event = match event {
-                        // Main events we are interested in.
-                        ToSwarm::GenerateEvent(event) => event,
-
-                        // Other events generated by the underlying behaviour are transparently
-                        // passed through.
-                        ToSwarm::Dial { opts } => {
-                            if opts.get_peer_id().is_none() {
-                                error!(
-                                    "The request-response isn't supposed to start dialing \
+        // We want to poll all the request-response protocols once, every time.
+        for (protocol, (behaviour, response_builder)) in &mut self.protocols {
+            if let Poll::Ready(event) = behaviour.poll(cx) {
+                // We almost always return Poll::Ready here, and we might have more work to do.
```

Comment:
I'll reformat this comment when I push my next set of revisions, after we do some initial testing.

```suggestion
                // We didn't receive Poll::Pending, so we might have more work to do.
                // And sometimes we add more work for ourselves via self.message_request,
                // so we need to schedule our own wakeup to make sure it gets done.
                // We usually return inside this block,
```